### PR TITLE
[libc++][NFC] Make AssertionInfoMatcher::CheckMessageMatches Stricter

### PR DESCRIPTION
**Pull request by hawkinsw**
Rather than allow for a message to be considered a match for the actual assertion if it is anywhere in the assertion text, make sure that the expected and the actual assertion are identical.

Addresses #77701

### CODE REVIEW

**Comment by hawkinsw** [comment]
Please feel free to ignore this change if it is unwanted. The tests that I ran locally all seem to pass with the stricter checking. I'll wait to see what it looks like on a full run of CI.

File: llvm/cmake/modules/LLVMExternalProjectUtils.cmake [hunk 1]
```diff
@@ -349,6 +349,8 @@ function(llvm_ExternalProject_Add name source_dir)
                -DPACKAGE_VERSION=${PACKAGE_VERSION}
                -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
                -DCMAKE_MAKE_PROGRAM=${CMAKE_MAKE_PROGRAM}
+               -DLLVM_ENABLE_SPHINX=${LLVM_ENABLE_SPHINX}
+               -DLLVM_BUILD_DOCS=${LLVM_BUILD_DOCS}
```
Comment (mordante) [inline llvm/cmake/modules/LLVMExternalProjectUtils.cmake:353]: Why are these changes needed? If needed, I think it would be better to do this in a separate review.

File: libcxx/test/support/check_assertion.h [hunk 2]
```diff
@@ -89,8 +89,7 @@ struct AssertionInfoMatcher {
     std::size_t found_at = got_msg.find(msg_);
     if (found_at == std::string_view::npos)
       return false;
-    // Allow any match
-    return true;
+    return found_at == 0 && got_msg.size() == msg_.size();
```
Comment (mordante) [inline libcxx/test/support/check_assertion.h:92]: I like this!

**Review by mordante — CHANGES_REQUESTED** [review]
In general I like this patch. However there seems to be unrelated changes.

**Review by ldionne — APPROVED** [review]
I am fine with this change if it passes CI (and without the unrelated changes). I thought we took advantage of this "hand wavyness" in more places, but if we don't and the CI is happy, I think this is an improvement.

File: llvm/cmake/modules/LLVMExternalProjectUtils.cmake (hunk 1 above)
Comment (hawkinsw) [reply llvm/cmake/modules/LLVMExternalProjectUtils.cmake:353]: Dang! I am clearly `git` stupid! Yes, these are unrelated. I will remove! Sorry!

File: libcxx/test/support/check_assertion.h (hunk 2 above)
Comment (hawkinsw) [reply libcxx/test/support/check_assertion.h:92]: Thanks!!

**Comment by hawkinsw** [comment]
Asking for another review from @mordante just out of caution!

**Comment by hawkinsw** [comment]
Pinging @var-const given his work on https://github.com/llvm/llvm-project/pull/77883. I will wait for his "go" before moving forward.

**Review by mordante — APPROVED** [review]
Thanks LGTM! Let's ship it.
